File: Algorithms/Java/ZigZag Conversion.java
public class Solution {
    public String convert(String s, int nRows) {
		if (s == null || nRows < 1) {
			return null;
		}
		if (nRows == 1) {
			return s;
		}
		StringBuilder[] sb = new StringBuilder[nRows];
		for (int i = 0; i < nRows; i++) {
			sb[i] = new StringBuilder();
		}
		boolean flag = false; // false stands for down and true for up
		int pos = 0; // Current position
		for (char c : s.toCharArray()) {
			sb[pos].append(String.valueOf(c));
			if (pos == nRows - 1) {
				flag = true;
			} else if (pos == 0) {
				flag = false;
			}
			if (flag) {
				pos--;
			} else {
				pos++;
			}
		}
		StringBuilder result = sb[0];
		for (int k = 1; k < nRows; k++) {
			result.append(sb[k]);
		}
		return result.toString();

	}
}

File: Algorithms/Python/ZigZag-Conversion.py
class Solution(object):

    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows == 1:
            return s
        ans = [''] * numRows
        row, flag = 0, True  # True for down, False for up
        for i in range(len(s)):
            ans[row] += s[i]
            if row == 0:
                flag = True
            elif row == numRows - 1:
                flag = False
            row = row + 1 if flag else row - 1
        return ''.join(ans)
